File: backend/app/sparql_builder.py
"""SPARQL generation from SHACL property specs plus the active filters."""
from typing import Any, Dict, List, Optional

from .namespaces import PREFIX_MAP, SPARQL_PREFIXES, ITEM_SEP, FIELD_SEP


def to_prefixed(iri: str) -> str:
    """Convert a full IRI to a SPARQL prefixed name (e.g. compass:country)."""
    for ns, prefix in PREFIX_MAP.items():
        if iri.startswith(ns):
            return prefix + iri[len(ns):]
    return f"<{iri}>"


def build_optional(spec: dict, lang: str) -> str:
    sid = spec["id"]
    path = to_prefixed(spec["path_iri"])
    cat = spec["category"]

    if cat == "lang_literal":
        return f'OPTIONAL {{ ?s {path} ?{sid} . FILTER(lang(?{sid}) = "{lang}") }}'
    if cat in ("simple_literal", "uri_literal", "boolean"):
        return f'OPTIONAL {{ ?s {path} ?{sid} . }}'
    if cat == "iri_with_label":
        return (
            f'OPTIONAL {{\n'
            f'            ?s {path} ?{sid}Node .\n'
            f'            OPTIONAL {{ ?{sid}Node skos:prefLabel ?{sid}Skos . FILTER(lang(?{sid}Skos) = "{lang}") }}\n'
            f'            OPTIONAL {{ ?{sid}Node rdfs:label ?{sid}Rdfs . FILTER(lang(?{sid}Rdfs) = "{lang}") }}\n'
            f'            BIND(COALESCE(?{sid}Skos, ?{sid}Rdfs) AS ?{sid}Lab)\n'
            f'        }}'
        )
    return ""


def build_select_expr(spec: dict) -> str:
    """GROUP_CONCAT for multi-valued properties, SAMPLE for single-valued ones."""
    sid = spec["id"]
    cat = spec["category"]
    is_multi = spec["is_multi"]

    if cat == "iri_with_label":
        if is_multi:
            return (
                f'(GROUP_CONCAT(DISTINCT CONCAT(STR(?{sid}Node), "{FIELD_SEP}", '
                f'COALESCE(?{sid}Lab, "")); separator="{ITEM_SEP}") AS ?{sid}Raw)'
            )
        return (
            f'(SAMPLE(?{sid}Node) AS ?{sid}Iri)\n'
            f'           (SAMPLE(?{sid}Lab) AS ?{sid}Label)'
        )
    if is_multi:
        return f'(GROUP_CONCAT(DISTINCT ?{sid}; separator="{ITEM_SEP}") AS ?{sid}Raw)'
    return f'(SAMPLE(?{sid}) AS ?{sid}Result)'


def _sparql_preamble(lang: str) -> str:
    """Type resolution, geometry and label binding, shared by both queries.

    The four entity classes carry coordinates and a compass:name. Country/Area
    concepts are additionally surfaced as map *regions*: they have no coordinates
    and label themselves with skos:prefLabel, so geometry is OPTIONAL and the
    label is COALESCEd across both naming properties.
    """
    return f"""
        {{
            ?s a compass:InternationalForum .
            BIND(compass:InternationalForum AS ?type)
        }} UNION {{
            ?s a compass:Network .
            BIND(compass:Network AS ?type)
        }} UNION {{
            ?s a compass:Project .
            BIND(compass:Project AS ?type)
        }} UNION {{
            ?s a compass:PartnerOrganization .
            BIND(compass:PartnerOrganization AS ?type)
        }} UNION {{
            ?s a compass:CountryArea .
            BIND(compass:CountryArea AS ?type)
        }}
        OPTIONAL {{ ?s geo:lat ?lat . }}
        OPTIONAL {{ ?s geo:long ?long . }}
        OPTIONAL {{ ?s compass:name ?nameLabel . FILTER(lang(?nameLabel) = "{lang}") }}
        OPTIONAL {{ ?s skos:prefLabel ?prefLabel . FILTER(lang(?prefLabel) = "{lang}") }}
        BIND(COALESCE(?nameLabel, ?prefLabel) AS ?label)
        FILTER(BOUND(?label))
        OPTIONAL {{ ?type rdfs:label ?typeLabel . FILTER(lang(?typeLabel) = "{lang}") }}
"""


def _special_optionals() -> str:
    """Properties fetched for display that no entity NodeShape declares."""
    return """
        OPTIONAL { ?s compass:startDate ?selfStart . }
        OPTIONAL { ?s compass:endDate ?selfEnd . }
        OPTIONAL { ?s compass:wpEntityTagIdEn ?wpEntityTagIdEn . }
        OPTIONAL { ?s compass:wpEntityTagIdDe ?wpEntityTagIdDe . }
"""


def _special_selects() -> str:
    return (
        '           (SAMPLE(?selfStart) AS ?selfStart)\n'
        '           (SAMPLE(?selfEnd) AS ?selfEnd)\n'
        '           (SAMPLE(?wpEntityTagIdEn) AS ?wpEntityTagIdEn)\n'
        '           (SAMPLE(?wpEntityTagIdDe) AS ?wpEntityTagIdDe)\n'
    )


def _union_or_single(parts: List[str]) -> str:
    if len(parts) > 1:
        return "{ " + " } UNION { ".join(parts) + " }"
    return parts[0]


def _build_where_clauses(
    query_params,
    filter_map: Dict[str, str],
    range_filters: Dict[str, str],
    date_filters: Dict[str, str],
    exclude_key: Optional[str] = None,
) -> List[str]:
    """exclude_key drops that dimension's own constraints, so facet counts for a
    dimension are not shrunk by the selection within it (drill-down faceting)."""
    where_clauses = []

    for key, val in query_params.items():
        if key == "lang" or key == exclude_key or not val:
            continue
        values = query_params.getlist(key)

        if key in filter_map:
            prop = filter_map[key]
            parts = []
            for v in values:
                if v.startswith("http") and ">" not in v:
                    parts.append(f"?s {prop} <{v}> .")
                else:
                    safe_v = v.replace('\\', '\\\\').replace('"', '\\"')
                    parts.append(f'?s {prop} ?{key}Val . FILTER(str(?{key}Val) = "{safe_v}")')
            if parts:
                where_clauses.append(_union_or_single(parts))

        elif key in date_filters:
            safe_v = val.replace('\\', '\\\\').replace('"', '\\"')
            prop = date_filters[key]
            where_clauses.append(
                f'OPTIONAL {{ ?s {prop} ?{key}Val . }} '
                f'FILTER(!BOUND(?{key}Val) || ?{key}Val >= "{safe_v}"^^xsd:date)'
            )

        elif key == "entityType":
            iri_list = ", ".join(
                f"<{v}>" for v in values if v.startswith("http") and ">" not in v
            )
            if iri_list:
                # Regions (CountryArea) are a separate visual layer and stay
                # visible regardless of the entity-type (pin) filter.
                where_clauses.append(
                    f"FILTER(?type IN ({iri_list}) || ?type = compass:CountryArea)"
                )

        elif key in range_filters:
            prop, datatype = range_filters[key]
            try:
                numeric_val = float(val)
                if datatype and "gYear" in datatype:
                    year_int = int(numeric_val)
                    where_clauses.append(
                        f'OPTIONAL {{ ?s {prop} ?{key}Val . }} '
                        f'FILTER(!BOUND(?{key}Val) || ?{key}Val >= "{year_int}"^^xsd:gYear)'
                    )
                else:
                    where_clauses.append(
                        f'OPTIONAL {{ ?s {prop} ?{key}Val . }} '
                        f'FILTER(!BOUND(?{key}Val) || ?{key}Val >= {numeric_val})'
                    )
            except ValueError:
                continue

    return where_clauses


def _categorize_specs(specs: List[Dict[str, Any]]):
    filter_map: Dict[str, str] = {}
    range_filters: Dict[str, str] = {}
    date_filters: Dict[str, str] = {}
    for spec in specs:
        prefixed = to_prefixed(spec["path_iri"])
        if spec["filter_type"] in ("multiselect", "toggle"):
            filter_map[spec["id"]] = prefixed
        elif spec["filter_type"] == "slider":
            range_filters[spec["id"]] = (prefixed, spec.get("datatype"))
        elif spec["filter_type"] == "datepicker":
            date_filters[spec["id"]] = prefixed
    return filter_map, range_filters, date_filters


def build_facet_query(
    specs: List[Dict[str, Any]], lang: str, query_params, target_id: str
) -> str:
    """Count entities per value of one tag dimension.

    The preamble is shared with build_entities_query so ?s ranges over exactly
    the same entity set the map renders.
    """
    filter_map, range_filters, date_filters = _categorize_specs(specs)
    target_path = filter_map[target_id]

    preamble = _sparql_preamble(lang)
    where_clauses = _build_where_clauses(
        query_params, filter_map, range_filters, date_filters, exclude_key=target_id
    )

    # Country/Area regions are background context, not results (see the map's
    # result badge, which counts point features only) — exclude them so facet
    # counts match what the map reports.
    sparql_where = preamble + f"        ?s {target_path} ?val .\n"
    sparql_where += "        FILTER(?type != compass:CountryArea)\n"
    if where_clauses:
        sparql_where += "        " + "\n        ".join(where_clauses) + "\n"

    return (
        SPARQL_PREFIXES
        + "    SELECT ?val (COUNT(DISTINCT ?s) AS ?n)\n"
        + "    WHERE {\n" + sparql_where + "    }\n"
        + "    GROUP BY ?val\n"
    )


def build_entities_query(
    specs: List[Dict[str, Any]], lang: str, query_params
) -> str:
    filter_map, range_filters, date_filters = _categorize_specs(specs)

    preamble = _sparql_preamble(lang)
    auto_optionals = "\n        ".join(build_optional(spec, lang) for spec in specs)
    auto_selects = "\n           ".join(build_select_expr(spec) for spec in specs)
    where_clauses = _build_where_clauses(query_params, filter_map, range_filters, date_filters)

    sparql_where = preamble + "        " + auto_optionals + "\n" + _special_optionals()
    if where_clauses:
        sparql_where += "        " + "\n        ".join(where_clauses) + "\n"

    return (
        SPARQL_PREFIXES
        + "    SELECT ?s ?label ?lat ?long ?type\n"
        + "           (SAMPLE(?typeLabel) AS ?typeLabelResult)\n"
        + "           " + auto_selects + "\n"
        + _special_selects()
        + "    WHERE {\n" + sparql_where + "    }\n"
        + "    GROUP BY ?s ?label ?lat ?long ?type\n"
    )

File: src/frontend/src/engine/sparqlBuilder.ts
/** SPARQL generation from the SHACL property specs plus the active filters. */
import { PREFIX_MAP, SPARQL_PREFIXES, ITEM_SEP, FIELD_SEP, type Spec, type Filters } from './namespaces';

/** dimension id -> selected values (each dimension applied once). */
export type ParamMap = Map<string, string[]>;

export function toParamMap(filters: Filters): ParamMap {
  const m: ParamMap = new Map();
  for (const [key, val] of Object.entries(filters)) {
    if (val === undefined) continue;
    const values = (Array.isArray(val) ? val : [val]).filter((v) => v !== undefined && v !== '');
    if (values.length) m.set(key, values.map(String));
  }
  return m;
}

/** Convert a full IRI to a SPARQL prefixed name (e.g. compass:country). */
export function toPrefixed(iri: string): string {
  for (const [ns, prefix] of Object.entries(PREFIX_MAP)) {
    if (iri.startsWith(ns)) return prefix + iri.slice(ns.length);
  }
  return `<${iri}>`;
}

export function buildOptional(spec: Spec, lang: string): string {
  const sid = spec.id;
  const path = toPrefixed(spec.path_iri);
  const cat = spec.category;

  if (cat === 'lang_literal') {
    return `OPTIONAL { ?s ${path} ?${sid} . FILTER(lang(?${sid}) = "${lang}") }`;
  }
  if (cat === 'simple_literal' || cat === 'uri_literal' || cat === 'boolean') {
    return `OPTIONAL { ?s ${path} ?${sid} . }`;
  }
  if (cat === 'iri_with_label') {
    return (
      `OPTIONAL {\n` +
      `            ?s ${path} ?${sid}Node .\n` +
      `            OPTIONAL { ?${sid}Node skos:prefLabel ?${sid}Skos . FILTER(lang(?${sid}Skos) = "${lang}") }\n` +
      `            OPTIONAL { ?${sid}Node rdfs:label ?${sid}Rdfs . FILTER(lang(?${sid}Rdfs) = "${lang}") }\n` +
      `            BIND(COALESCE(?${sid}Skos, ?${sid}Rdfs) AS ?${sid}Lab)\n` +
      `        }`
    );
  }
  return '';
}

/** GROUP_CONCAT for multi-valued properties, SAMPLE for single-valued ones. */
export function buildSelectExpr(spec: Spec): string {
  const sid = spec.id;
  const cat = spec.category;
  const isMulti = spec.is_multi;

  if (cat === 'iri_with_label') {
    if (isMulti) {
      return (
        `(GROUP_CONCAT(DISTINCT CONCAT(STR(?${sid}Node), "${FIELD_SEP}", ` +
        `COALESCE(?${sid}Lab, "")); separator="${ITEM_SEP}") AS ?${sid}Raw)`
      );
    }
    return `(SAMPLE(?${sid}Node) AS ?${sid}Iri)\n           (SAMPLE(?${sid}Lab) AS ?${sid}Label)`;
  }
  if (isMulti) {
    return `(GROUP_CONCAT(DISTINCT ?${sid}; separator="${ITEM_SEP}") AS ?${sid}Raw)`;
  }
  return `(SAMPLE(?${sid}) AS ?${sid}Result)`;
}

const PIN_CLASSES = ['InternationalForum', 'Network', 'Project', 'PartnerOrganization'] as const;

/**
 * Which variable a set of filter clauses constrains.
 *
 * Filters read the same for a pin the map draws and for the pin that puts a
 * region on the map, but they cannot share variable names: the region branch
 * nests its copy inside FILTER EXISTS, where the outer ?type is already bound
 * to compass:CountryArea.
 */
type Subject = {
  var: string;
  typeVar: string;
  /** Keeps helper variables distinct across the two copies. */
  suffix: string;
  /** Bind typeVar here, rather than relying on an outer BIND. */
  declareType: boolean;
};

const PIN: Subject = { var: '?s', typeVar: '?type', suffix: '', declareType: false };
const REGION_PIN: Subject = { var: '?pin', typeVar: '?pinType', suffix: 'Pin', declareType: true };

/** The four entity classes that carry coordinates -- the pins on the map. */
function pinBranch(whereClauses: string[], indent = '        '): string {
  const branches = PIN_CLASSES.map(
    (name) => `{ ?s a compass:${name} . BIND(compass:${name} AS ?type) }`,
  ).join(`\n${indent}UNION `);
  let body = `${indent}${branches}\n`;
  if (whereClauses.length) body += indent + whereClauses.join(`\n${indent}`) + '\n';
  return body;
}

/**
 * Country/Area concepts, reachable only through a pin that points at one.
 *
 * A region is a shaded polygon rather than a result, and it carries no tags of
 * its own: it reaches the map because some pin passing the same filters records
 * it, so shading always means "matching pins are in here".
 */
function regionBranch(whereClauses: string[], indent = '        '): string {
  let inner = `${indent}    ?pin compass:countryArea ?s .\n`;
  if (whereClauses.length) {
    inner += `${indent}    ` + whereClauses.join(`\n${indent}    `) + '\n';
  }
  return (
    `${indent}?s a compass:CountryArea .\n` +
    `${indent}BIND(compass:CountryArea AS ?type)\n` +
    `${indent}FILTER EXISTS {\n${inner}${indent}}\n`
  );
}

/**
 * Geometry and label binding, applied to pins and regions alike.
 *
 * Regions have no coordinates and label themselves with skos:prefLabel, so
 * geometry is OPTIONAL and the label is COALESCEd across both properties.
 */
function sharedOptionals(lang: string): string {
  return `        OPTIONAL { ?s geo:lat ?lat . }
        OPTIONAL { ?s geo:long ?long . }
        OPTIONAL { ?s compass:name ?nameLabel . FILTER(lang(?nameLabel) = "${lang}") }
        OPTIONAL { ?s skos:prefLabel ?prefLabel . FILTER(lang(?prefLabel) = "${lang}") }
        BIND(COALESCE(?nameLabel, ?prefLabel) AS ?label)
        FILTER(BOUND(?label))
        OPTIONAL { ?type rdfs:label ?typeLabel . FILTER(lang(?typeLabel) = "${lang}") }
`;
}

function specialOptionals(): string {
  return `
        OPTIONAL { ?s compass:startDate ?selfStart . }
        OPTIONAL { ?s compass:endDate ?selfEnd . }
        OPTIONAL { ?s compass:wpEntityTagId ?wpEntityTagId . }
`;
}

function specialSelects(): string {
  return (
    '           (SAMPLE(?selfStart) AS ?selfStart)\n' +
    '           (SAMPLE(?selfEnd) AS ?selfEnd)\n' +
    '           (SAMPLE(?wpEntityTagId) AS ?wpEntityTagId)\n'
  );
}

function unionOrSingle(parts: string[]): string {
  if (parts.length > 1) return '{ ' + parts.join(' } UNION { ') + ' }';
  return parts[0];
}

function escapeLiteral(v: string): string {
  return v.replace(/\\/g, '\\\\').replace(/"/g, '\\"');
}

function buildWhereClauses(
  params: ParamMap,
  filterMap: Map<string, string>,
  rangeFilters: Map<string, [string, string | null]>,
  dateFilters: Map<string, string>,
  subject: Subject = PIN,
  excludeKey: string | null = null,
): string[] {
  const whereClauses: string[] = [];
  const subj = subject.var;

  for (const [key, values] of params) {
    if (key === 'lang' || key === excludeKey || values.length === 0) continue;
    const v0 = `?${key}${subject.suffix}Val`;

    if (filterMap.has(key)) {
      const prop = filterMap.get(key)!;
      const parts: string[] = [];
      for (const v of values) {
        if (v.startsWith('http') && !v.includes('>')) {
          parts.push(`${subj} ${prop} <${v}> .`);
        } else {
          parts.push(`${subj} ${prop} ${v0} . FILTER(str(${v0}) = "${escapeLiteral(v)}")`);
        }
      }
      if (parts.length) whereClauses.push(unionOrSingle(parts));
    } else if (dateFilters.has(key)) {
      const prop = dateFilters.get(key)!;
      const safeV = escapeLiteral(values[0]);
      whereClauses.push(
        `OPTIONAL { ${subj} ${prop} ${v0} . } ` +
          `FILTER(!BOUND(${v0}) || ${v0} >= "${safeV}"^^xsd:date)`,
      );
    } else if (key === 'entityType') {
      const iriList = values
        .filter((v) => v.startsWith('http') && !v.includes('>'))
        .map((v) => `<${v}>`)
        .join(', ');
      if (iriList) {
        // A region has no type of its own to filter, so the legend reaches it
        // through the pins: hide every Project and a region holding only
        // projects stops being shaded.
        const filter = `FILTER(${subject.typeVar} IN (${iriList}))`;
        whereClauses.push(
          subject.declareType ? `${subj} a ${subject.typeVar} . ${filter}` : filter,
        );
      }
    } else if (rangeFilters.has(key)) {
      const [prop, datatype] = rangeFilters.get(key)!;
      const numericVal = Number(values[0]);
      if (Number.isNaN(numericVal)) continue;
      if (datatype && datatype.includes('gYear')) {
        const yearInt = Math.trunc(numericVal);
        whereClauses.push(
          `OPTIONAL { ${subj} ${prop} ${v0} . } ` +
            `FILTER(!BOUND(${v0}) || ${v0} >= "${yearInt}"^^xsd:gYear)`,
        );
      } else {
        whereClauses.push(
          `OPTIONAL { ${subj} ${prop} ${v0} . } ` +
            `FILTER(!BOUND(${v0}) || ${v0} >= ${numericVal})`,
        );
      }
    }
  }

  return whereClauses;
}

function categorizeSpecs(specs: Spec[]) {
  const filterMap = new Map<string, string>();
  const rangeFilters = new Map<string, [string, string | null]>();
  const dateFilters = new Map<string, string>();
  for (const spec of specs) {
    const prefixed = toPrefixed(spec.path_iri);
    if (spec.filter_type === 'multiselect' || spec.filter_type === 'toggle') {
      filterMap.set(spec.id, prefixed);
    } else if (spec.filter_type === 'slider') {
      rangeFilters.set(spec.id, [prefixed, spec.datatype]);
    } else if (spec.filter_type === 'datepicker') {
      dateFilters.set(spec.id, prefixed);
    }
  }
  return { filterMap, rangeFilters, dateFilters };
}

/**
 * Count-per-value query for one tag dimension (drill-down faceting).
 *
 * Regions are background context rather than results (see the map's result
 * badge, which counts point features only), so only the pin branch is counted.
 */
export function buildFacetQuery(specs: Spec[], lang: string, params: ParamMap, targetId: string): string {
  const { filterMap, rangeFilters, dateFilters } = categorizeSpecs(specs);
  const targetPath = filterMap.get(targetId)!;

  const whereClauses = buildWhereClauses(params, filterMap, rangeFilters, dateFilters, PIN, targetId);

  let sparqlWhere = pinBranch(whereClauses);
  sparqlWhere += `        ?s ${targetPath} ?val .\n`;
  sparqlWhere += sharedOptionals(lang);

  return (
    SPARQL_PREFIXES +
    '    SELECT ?val (COUNT(DISTINCT ?s) AS ?n)\n' +
    '    WHERE {\n' +
    sparqlWhere +
    '    }\n' +
    '    GROUP BY ?val\n'
  );
}

/** The full SPARQL SELECT query for the entities endpoint. */
export function buildEntitiesQuery(specs: Spec[], lang: string, params: ParamMap): string {
  const { filterMap, rangeFilters, dateFilters } = categorizeSpecs(specs);

  const autoOptionals = specs.map((spec) => buildOptional(spec, lang)).join('\n        ');
  const autoSelects = specs.map((spec) => buildSelectExpr(spec)).join('\n           ');
  const pinClauses = buildWhereClauses(params, filterMap, rangeFilters, dateFilters, PIN);
  const regionClauses = buildWhereClauses(params, filterMap, rangeFilters, dateFilters, REGION_PIN);

  let sparqlWhere =
    '        {\n' +
    pinBranch(pinClauses, '            ') +
    '        } UNION {\n' +
    regionBranch(regionClauses, '            ') +
    '        }\n';
  sparqlWhere += sharedOptionals(lang);
  sparqlWhere += '        ' + autoOptionals + '\n' + specialOptionals();

  return (
    SPARQL_PREFIXES +
    '    SELECT ?s ?label ?lat ?long ?type\n' +
    '           (SAMPLE(?typeLabel) AS ?typeLabelResult)\n' +
    '           ' +
    autoSelects +
    '\n' +
    specialSelects() +
    '    WHERE {\n' +
    sparqlWhere +
    '    }\n' +
    '    GROUP BY ?s ?label ?lat ?long ?type\n'
  );
}
